sirepo/job_supervisor.py

```diff
@@ -861,31 +956,98 @@ async def _receive_api_runSimulation(self, req, recursing=False):
-            if not r:
-                raise AssertionError(f"no reply to req={req}")
-            o.run_callback = tornado.ioloop.IOLoop.current().call_later(
-                0,
-                self._run,
-                op=o,
-                curr_db=self.db,
-                prev_db=d,
+                self._db_file(self.db.computeJid).purebasename
             )
-            o = None
+            return t
+
+        async def _valid_or_reply(force_run):
+            if self._is_running_pending():
+                if not self._req_is_valid(req):
+                    return PKDict(
+                        state=job.ERROR,
+                        error="another browser is running the simulation",
+                    )
+                if force_run:
+                    return PKDict(
+                        state=job.ERROR,
+                        error="simulation state unknown, refresh the browser",
+                    )
+                # Not _receive_api_runStatus, because runStatus should have been
+                # called before this function is called.
+                return self._status_reply(req)
+            if (
+                not force_run
+                and self._req_is_valid(req)
+                and self.db.status == job.COMPLETED
+            ):
+                # TODO(robnagler) simplify after https://github.com/radiasoft/sirepo/issues/7386
+
+                # Valid, completed, sequential simulation
+                # Read this first https://github.com/radiasoft/sirepo/issues/2007
+                r = await self._receive_api_runStatus(req)
+                if r.state == job.MISSING:
+                    # happens when the run dir is deleted (ex purge_non_premium)
+                    if recursing:
+                        raise AssertionError(f"already called from self req={req}")
+                    # Rerun the simulation, since there's no "button" in the UI for
+                    # this case.
+                    return await self._receive_api_runSimulation(req, recursing=True)
+                return r
+            return None
+
+        if self._run_status_op:
+            raise AssertionError(f"_run_status_op already set job={self}")
+        # Reply case yields, but does not modify global state
+        if r := await _valid_or_reply(req.content.data.get("forceRun")):
             return r
+        # TODO(robnagler) consolidate _start_run_status_op
+        req.content.computeJobSerial = _update_db()
+        d = self.db.dbUpdateTime
+        try:
+            r = await self._send_with_reply(
+                job.OP_RUN,
+                req,
+                is_run_status_op=True,
+                jobCmd=job.CMD_COMPUTE,
+                nextRequestSeconds=self.db.nextRequestSeconds,
+            )
         except Exception as e:
-            if o:
-                o.destroy(internal_error=f"_receive_api_runSimulation exception={e}")
+            if self.db.dbUpdateTime == d:
```

Comment:
I'm concerned about this code. If a job needs to wait on another simulation to complete (op_slot.alloc or cpu_slot.alloc) the dbUpdateTime is updated (_db_write_file is called when setting job situation). If it then raises at some point in the send I think we want the status to be set to job.ERROR but it won't be because the conditional is not True. I don't exactly know what causes a raise in send so maybe it is unlikely enough to not be worth covering. I think the job will be "missing" which is probably ok. 

A larger concern is that the code could be changed in some way in the future to where the dbupdatetime is modified during the normal course of events. It is always modified during _db_update which seems like a change someone could easily introduce. I don't think it will be obvious to the person changing the code that this case is now broken. Not sure how to protect other than possibly using some other field besides dbUpdateTime for the check.